### PR TITLE
nixos/netbird: fix port conflict on metrics endpoint

### PR DESCRIPTION
**Pull request by TheRealGramdalf**
Added changes discussed in https://github.com/NixOS/nixpkgs/pull/354032#issuecomment-2480925927 to fix a port clash between the netbird management and signal services, as well as an `extraOption` to facilitate ad-hoc changes in the future should more breaking changes be released by upstream.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [X] (Package updates) Added a release notes entry if the change is major or breaking
  - [X] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by TheRealGramdalf** [comment]
I based this on the netbird 0.32.0 release before realizing that PR was merged to master after the branch off - it would be helpful to know from @Saturn745 or @vrifox whether I should rebase this PR against the 0.31.1 release currently in the branch off (which still has the issues fixed) or if the 0.32.0 release PR should be backported as well.

File: nixos/modules/services/networking/netbird/signal.nix
Comment (PatrickDaG) [inline nixos/modules/services/networking/netbird/signal.nix:92]: I think you have to do brackets around this as the call has higher precedence.

**Review by PatrickDaG — CHANGES_REQUESTED** [review]

File: nixos/modules/services/networking/netbird/signal.nix
Comment (SuperSandro2000) [reply nixos/modules/services/networking/netbird/signal.nix:51]:
```suggestion
      default = [ "--metrics-port" "9091" ];
```

File: nixos/modules/services/networking/netbird/signal.nix
Comment (SuperSandro2000) [inline nixos/modules/services/networking/netbird/signal.nix:88]:
```suggestion
            "--port"
            cfg.port
            "--log-file"
            "console"
```

**Review by oddlama — COMMENTED** [review]
> I based this on the netbird 0.32.0 release before realizing that PR was merged to master after the branch off - it would be helpful to know from @Saturn745 or @vrifox whether I should rebase this PR against the 0.31.1 release currently in the branch off (which still has the issues fixed) or if the 0.32.0 release PR should be backported as well.

No need, backporting will only consider changes in this PR so it doesn't matter what you base this against

> which still has the issues fixed

What do you mean? Assuming you mean the issue is not present in 0.31.1 then I guess we won't have to backport at all

File: nixos/doc/manual/release-notes/rl-2411.section.md
Comment (oddlama) [reply nixos/doc/manual/release-notes/rl-2411.section.md:39]: Not sure how backporting affects release notes, but I would presume that keeping the notes will require a manual backport

File: nixos/modules/services/networking/netbird/signal.nix
Comment (PatrickDaG) [reply nixos/modules/services/networking/netbird/signal.nix:51]: I've thought about it and looked at the `escapeSystemdExecArgs` method, and @SuperSandro2000 is probably right we should leave it inside the call and just quote it ourselves, maybe add an example and/or a note so people know they have to do this.

Comment (TheRealGramdalf) [reply nixos/modules/services/networking/netbird/signal.nix:51]: Alternatively I could join it with `=` to keep it as a single string without whitespaces, though splitting it into two strings is consistent with the module internally

File: nixos/doc/manual/release-notes/rl-2411.section.md
Comment (TheRealGramdalf) [reply nixos/doc/manual/release-notes/rl-2411.section.md:39]: I asked on the matrix channel and they said this should definitely be backported, so I'm guessing the release note should stay for 24.11

**Comment by TheRealGramdalf** [comment]
> What do you mean? Assuming you mean the issue is not present in 0.31.1 then I guess we won't have to backport at all

Both 0.31.1 and 0.32.0 are affected by the port clash. I realized after I created the PR that the 0.32.0 update PR didn't make it into the branch off.
This only really matters because of the release notes, since theres a link in there pointing to the 0.32.0 release notes. My question was whether or not I should change the link to point at 0.31.1, or if we should backport the 0.32.0 PR as well.

File: nixos/modules/services/networking/netbird/signal.nix
Comment (TheRealGramdalf) [reply nixos/modules/services/networking/netbird/signal.nix:56]: Ah good catch, I accidentally switched which port I changed compared to my snippet

**Comment by TheRealGramdalf** [comment]
> I think if we're backporting this we should do it right and add a dedicated `metricsPort` option, you can just copy mine from the other PR. Sounds like a bad idea to rely on a default `extraOption`, those should only be used in special cases.

Will do. I'll still add the `extraOptions` to signal to make it consistent and to aid in case of further breakages from upstream.

File: nixos/modules/services/networking/netbird/signal.nix
Comment (TheRealGramdalf) [reply nixos/modules/services/networking/netbird/signal.nix:88]: Any particular reason to remove the comments? I didn't touch them since the intent was for this to be a small, simple fix focused on the port clash, but I'm open to removing them if needed.

**Comment by TheRealGramdalf** [comment]
> Both 0.31.1 and 0.32.0 are affected by the port clash. I realized after I created the PR that the 0.32.0 update PR didn't make it into the branch off. This only really matters because of the release notes, since theres a link in there pointing to the 0.32.0 release notes. My question was whether or not I should change the link to point at 0.31.1, or if we should backport the 0.32.0 PR as well.

I'll just go ahead and change this to 0.31.1. If #356512 gets backported before this is merged I can switch it back, but for now I'll assume the version of netbird in the 24.11 release is going to be 0.31.1.

File: nixos/modules/services/networking/netbird/signal.nix
Comment (TheRealGramdalf) [reply nixos/modules/services/networking/netbird/signal.nix]: Since the metrics port is now in a dedicated option I opted to remove this note entirely

File: nixos/modules/services/networking/netbird/signal.nix
Comment (TheRealGramdalf) [reply nixos/modules/services/networking/netbird/signal.nix]: Same deal as the note, this is no longer necessary since it's using a dedicated option now

File: nixos/doc/manual/release-notes/rl-2411.section.md
Comment (TheRealGramdalf) [reply nixos/doc/manual/release-notes/rl-2411.section.md]: > Not sure how backporting affects release notes, but I would presume that keeping the notes will require a manual backport

I believe a manual backport would be necessary regardless, unless one of the package maintainers were to initiate it once this gets merged

**Review by PatrickDaG — APPROVED** [review]

**Review by oddlama — APPROVED** [review]

**Comment by TheRealGramdalf** [comment]
Ready for merge. If I need to create another PR manually for the backport just let me know.

**Comment by TheRealGramdalf** [comment]
Yeah, I'll wait until it's clear it won't be getting in, no need for extra effort evals. I was hoping to get it in for 24.11 not only for the port fix, but to make people aware of the additional ports in case of other conflicts/if updates to firewall rules are required. Ideally I'd have a note specifically mentioning that (upstream communication isn't the greatest; it looks like they test almost exclusively in docker where clashing ports aren't an issue), but this'll have to do for now.

File: nixos/modules/services/networking/netbird/management.nix
Comment (JohnRTitor) [inline nixos/modules/services/networking/netbird/management.nix:1]: Add an assert here, which assumes `port != metricsPort`

Comment (TheRealGramdalf) [reply nixos/modules/services/networking/netbird/management.nix:1]: So would this work?
```nix
assertions = [
      {
        assertion = cfg.port != cfg.metricsPort;
        message = "The primary listen port cannot be the same as the listen port for the metrics endpoint";
      }
    ];
```
I verified that it does syntactically (i.e. it correctly identifies and throws an error when I set them to the same value), I'm just not sure if you meant to use `assertions` or `assert`.

Comment (JohnRTitor) [reply nixos/modules/services/networking/netbird/management.nix:1]: Yes, just like this https://github.com/NixOS/nixpkgs/blob/69fd9db72856a01f3230e3eb66f7f532e2566c50/nixos/modules/services/scheduling/scx.nix#L101

**Comment by TheRealGramdalf** [comment]
Once again ready for merge. I did some manual checks and everything functions fine, not sure if we should wait for ofborg or just send it due to the branch window - I'm relatively new to PRs so I'm not too sure if it's required or just recommended.

**Comment by JohnRTitor** [comment]
Well we mainly wait for Ofborg so it detects any underlying issue during eval. But considering how slower and costlier it has become, there have been calls to replace it with GitHub actions, which should be faster. 

New eval workflow PR has been merged yesterday, and it looks like the checks passed. Let's still wait for Ofborg to check other things, I am happy to merge it after that.

**Comment by TheRealGramdalf** [comment]
Allright, sounds good! I'm mainly just trying to get it in before the 24.11 release so the change is in the release notes. I've been advised over on matrix that this should ideally get backported, but again I'm new to PRs and I'm not entirely sure what the process is like - there were mentions of changes to the release notes requiring a manual backport, but again I'm new to that specific area.

**Comment by TheRealGramdalf** [comment]
Checks complete, ready for merge - let me know if there's anything to do for the backport PR.

**Review by Scrumplex — APPROVED** [review]
Changes LGTM

**Pull request merged**
